### PR TITLE
swap: Refactor swap logs with actions prefix

### LINKED ISSUE #1788: Refactor swap logs with actions prefix
In order to filter swap logs, the prefix of swap logs has to be refactored

currently swap logs are prefixed with 

> swap-logs "*"

This creates a conflict when searching such logs.

as @skylenet mentioned we should change the logs to use actions instead of a generic prefix. 

> For now I would prefer to just tag your log line like:  swap.{action}  e.g. swap.contractDeployed   addr=1234 owner=4123. You can then easily filter the logs by those tags.

### PR DESCRIPTION
**Pull request by santicomp2014**
This PR aims to resolve #1788, which is about refactoring swap log-prefix with a more appropriate message.

This enables better indexing and searching for swap related logs.

Example instead of `swap-logs "*"` it should be `swap.contractDeployed`

### CODE REVIEW

**Review by mortelli — CHANGES_REQUESTED** [review]
good first iteration. 

the idea of setting the action in a field (rather than directly in the context) and calling each write with a new context that appends it is a good one, since we have no public methods to change the context for the logger once set.

that said, there are a few things to polish. some additional pointers:
- it would be nice for the PR to include a high level summary from a technical point of view, so that reviewers can understand the idea behind the implementation more easily.
- make sure all swap logs have an action set, at least the `checkChainID` function is without one.
- since the general logger is shared, please make sure there are no possible concurrency problems if 2+ loggers were to set an action consecutively and then call subsequent writes.
- remember to keep the branch up to date with `master`.

and as an opinion: i would like to see more consistency with how each log write is called. i think i favor the approach taken in the `cashCheque` function the best, which is setting the log action at the beginning of the function. the more i think of it, the more i like this idea rather than passing the action as part of the context along with other business fields. but this is certainly up for discussion.

File: log/log.go [hunk 1]
```diff
@@ -5,12 +5,27 @@ import (
 	"github.com/ethereum/go-ethereum/metrics"
 )
 
+// Action Defines swap log actions
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: I agree this was not the original idea, It makes for sense in swap as you mention.
I refactored log into swap package, this cleans up the code.

File: log/log.go [hunk 2]
```diff
@@ -5,12 +5,27 @@ import (
...
+type Action string
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: The solution I found was using a type alias for passing constant like variables.

The idea was to use always the same type of action, but since there are too many variations, I resolved to remove actions altogether.

File: log/log.go [hunk 3]
```diff
@@ -5,12 +5,27 @@ import (
...
+
 const (
 	// CallDepth is set to 1 in order to influence to reported line number of
 	// the log message with 1 skipped stack frame of calling l.Output()
 	CallDepth = 1
+	//DefaultAction is the default action filter for SwapLogs
+	DefaultAction Action = "*"
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: Resolved, I changed this to `"swap_action","undefined"`

File: swap/swap.go [hunk 4]
```diff
@@ -301,6 +303,7 @@ func (s *Swap) handleEmitChequeMsg(ctx context.Context, p *Peer, msg *EmitCheque
 	defer p.lock.Unlock()
 
 	cheque := msg.Cheque
+	p.logger.SetLogAction("handle_cheque")
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap.go:338]: Resolved

File: log/log.go [hunk 5]
```diff
@@ -47,6 +62,104 @@ func Trace(msg string, ctx ...interface{}) {
...
+	for _, elem := range ctx {
+		if elem == "action" && len(ctx)%2 == 0 {
+			return ctx
+		}
+	}
+	ctx = addSwapAction(sl, ctx...)
+	return ctx
+}
+
+// Warn TODO REVIEW THIS COMMENT is a convenient alias for log.Warn with stats
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: Fixed

File: log/log.go [hunk 6]
```diff
@@ -47,6 +62,104 @@ func Trace(msg string, ctx ...interface{}) {
...
+
+func addSwapCtx(sl SwapLogger, ctx ...interface{}) []interface{} {
+	ctx = append([]interface{}{"base", sl.overlayAddr}, ctx...)
+	if sl.peerID != "" {
+		ctx = append(ctx, "peer", sl.peerID)
+	}
+	return ctx
+}
+
+func addSwapAction(sl SwapLogger, ctx ...interface{}) []interface{} {
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: Removed

File: log/log.go [hunk 7]
```diff
@@ -47,6 +62,104 @@ func Trace(msg string, ctx ...interface{}) {
...
+		action:      DefaultAction,
+		overlayAddr: overlayAddr,
+		peerID:      peerID,
+	}
+	ctx := addSwapCtx(swapLogger)
+	swapLogger.logger = l.New(ctx...)
+	return swapLogger
+}
+
+func addSwapCtx(sl SwapLogger, ctx ...interface{}) []interface{} {
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: This was refactored, as ctx is not used in this function.
I removed `ctx` and introduced `overlayAddr,peerID`

File: log/log.go [hunk 8]
```diff
@@ -47,6 +62,104 @@ func Trace(msg string, ctx ...interface{}) {
...
+	if action == "" {
+		sl.action = DefaultAction
+		return
+	}
+	//Todo validate it's a specific action, if not default
+	sl.action = action
+}
+
+// NewSwapLogger is an alias for log.New
+func NewSwapLogger(overlayAddr string) (swapLogger SwapLogger) {
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: Refactored into a single function that returns a new Logger

File: log/log.go [hunk 9]
```diff
@@ -5,12 +5,27 @@ import (
...
+	//DefaultAction is the default action filter for SwapLogs
+	DefaultAction Action = "*"
+	//SentChequeAction is a filter for SwapLogs
+	SentChequeAction Action = "SentCheque"
 )
 
+// SwapLogger wraps the ethereum logger with specific information for swap logging
+type SwapLogger struct {
+	action      Action
+	overlayAddr string
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: Removed overlayAddr & peerID.
The struct also contains the logger instance.

When a new SwapLogger is initialized the logger is also initialized.
So it made sense to keep it inside the struct.

File: swap/log.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
+package swap
+
+import (
+	l "github.com/ethereum/go-ethereum/log"
```
Comment (mortelli) [inline swap/log.go]: in this case i also think we should use `log` instead of `l`, for clarity's sake

File: swap/log.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+		action: DefaultAction,
+	}
+	ctx := addSwapCtx(swapLogger, overlayAddr, peerID)
+	swapLogger.logger = l.New(ctx...)
+	return swapLogger
+}
+
+func addSwapCtx(sl Logger, overlayAddr string, peerID string) []interface{} {
+	ctx := []interface{}{"base", overlayAddr}
+	if peerID != "" {
```
Comment (mortelli) [inline swap/log.go]: as i said in my previous review, i'm not a fan of this style of defensive programming.

i'd rather we just called an extra line/function for the peer logger instead of passing an empty parameter.

just my 2¢

File: swap/log.go [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+		sl.action = DefaultAction
+		return
+	}
+	//Todo validate it's a specific action, if not default
+	sl.action = action
+}
+
+// NewSwapLogger is an alias for log.New
+func NewSwapLogger(overlayAddr string) (swapLogger Logger) {
+	return newLogger(overlayAddr, "")
```
Comment (mortelli) [inline swap/log.go]: as i mentioned, i would entirely remove this type of call (one with a `peerID` value of `""`) but if you are keeping it, just make `newLogger` public and call it with the appropriate parameters.

`NewSwapLogger` and `NewSwapPeerLogger` are just indirection in this current implementation.

File: swap/swap.go [hunk 13]
```diff
@@ -154,6 +154,7 @@ func New(dbPath string, prvkey *ecdsa.PrivateKey, backendURL string, params *Par
 	if backendURL == "" {
 		return nil, errors.New("no backend URL given")
 	}
+	swapLog.SetLogAction("swap_init")
```
Comment (mortelli) [inline swap/swap.go]: this will cause `swap` to appear twice in the log

`swap_action=swap_init`

File: swap/log.go [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+const (
+	// CallDepth is set to 1 in order to influence to reported line number of
+	// the log message with 1 skipped stack frame of calling l.Output()
+	CallDepth = 1
+	//DefaultAction is the default action filter for SwapLogs
+	DefaultAction string = "undefined"
+)
+
+// Logger wraps the ethereum logger with specific information for swap logging
+type Logger struct {
```
Comment (mortelli) [inline swap/log.go:15]: why was this renamed from `SwapLogger` to `Logger`?

**Review by mortelli — COMMENTED** [review]
we're not quite there yet, but the PR is moving in the right direction.

(as a side note: please make sure all previously raised issues have been resolved or addressed before requesting another review)

File: swap/log.go (hunk 14 above)
Comment (santicomp2014) [reply swap/log.go:15]: Redundant naming, the package is called swap and the field is SwapLogger, the linter complains of stutter code.

File: swap/log.go [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+const (
+	// CallDepth is set to 1 in order to influence to reported line number of
+	// the log message with 1 skipped stack frame of calling l.Output()
+	CallDepth = 1
+	//DefaultAction is the default action filter for SwapLogs
+	DefaultAction string = "undefined"
+)
+
+// Logger wraps the ethereum logger with specific information for swap logging
```
Comment (Eknir) [inline swap/log.go:12]: What is meant with `ethereum logger`

File: swap/log.go [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// SetLogAction set the current log action prefix
+func (sl *Logger) SetLogAction(action string) {
+	//Adds default action undefined
+	if action == "" {
+		sl.action = DefaultAction
+		return
+	}
+	//Todo validate it's a specific action, if not default
```
Comment (Eknir) [inline swap/log.go]: Do you intend to resolve this one before merging?
If you don't intend to solve it, could you be a bit more specific here?

File: swap/log.go (hunk 11 above)
Comment (Eknir) [reply swap/log.go]: It looks good to me. @mortelli , what is your problem with this?

Comment (mortelli) [reply swap/log.go]: i don't think it's good style.

why should a function allow an empty string as a peer ID? it is not conceptually correct; either you expect an ID or you don't. 

functions should be focused on doing one thing and doing it well, not trying to withstand whatever call is thrown at them. if a call is wrong, then let it fail. this `if` hides a wrong call, should it happen.

checking if `peerID` is empty means that both of these calls are correct:

`addSwapCtx(myLogger, myAddress, myPeerID)`
`addSwapCtx(myLogger, myAddress, "")`

[here](https://github.com/ethersphere/swarm/pull/1872#discussion_r334962257) is my previous comment about this.

File: swap/log.go [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+func newLogger(overlayAddr string, peerID string) (swapLogger Logger) {
+	swapLogger = Logger{
+		action: DefaultAction,
+	}
+	ctx := addSwapCtx(swapLogger, overlayAddr, peerID)
+	swapLogger.logger = log.New(ctx...)
+	return swapLogger
+}
+
+func addSwapCtx(sl Logger, overlayAddr string, peerID string) []interface{} {
```
Comment (mortelli) [inline swap/log.go]: why is this function called `add` if it in fact returns a context?

also: the `sl` parameter is not used

File: swap/log.go (hunk 15 above)
Comment (mortelli) [reply swap/log.go:12]: the comment could be clearer in terms of the logger being wrapped specifically with the swap `action`

File: swap/log.go (hunk 14 above)
Comment (mortelli) [reply swap/log.go:15]: that clears it up, thanks

File: swap/log.go (hunk 11 above)
Comment (mortelli) [reply swap/log.go]: here's how you can do it with 1 less function, no `if`s and no duplicated code:

```golang
// newSwapLogger returns a new swap logger with standard context
func newSwapLogger(logPath string, overlayAddr []byte) Logger {
	ctx := []interface{}{"base", overlayAddr}
	return newLogger(logPath, ctx)
}

// newPeerLogger returns a new swap logger with a context that includes peer ID
func newPeerLogger(s *Swap, peerID enode.ID) Logger {
	ctx := []interface{}{"base", s.params.OverlayAddr[:16], "peer", peerID}
	return newLogger(s.params.LogPath, ctx)
}

func newLogger(logPath string, ctx []interface{}) Logger {
	swapLogger := Logger{
		action: DefaultAction,
	}
	swapLogger.logger = log.New(ctx...)
	setLoggerHandler(logPath, swapLogger.GetLogger())
	return swapLogger
}
```

Comment (Eknir) [reply swap/log.go]: I like your proposal @mortelli . Having one function (`newLogger`) for both the SwapLogger and the PeerLogger is indeed confusing (in fact, it confused me). Your proposal makes it clear what kind of logger we are creating and makes the if statement redundant.

File: log/log.go [hunk 18]
```diff
@@ -47,6 +62,104 @@ func Trace(msg string, ctx ...interface{}) {
...
+		ctx = append(ctx, "peer", sl.peerID)
+	}
+	return ctx
+}
+
+func addSwapAction(sl SwapLogger, ctx ...interface{}) []interface{} {
+	return append([]interface{}{"action", sl.action}, ctx...)
+}
+
+// GetLogger return the underlining logger
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: Fixed

File: swap/swap.go [hunk 19]
```diff
@@ -35,19 +35,20 @@ import (
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/ethclient"
-	"github.com/ethereum/go-ethereum/log"
+	l "github.com/ethereum/go-ethereum/log"
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap.go]: I think maybe swarmLog is a more appropriate name for this logger.
The loggers  are 

- swarmLog --> github.com/ethereum/go-ethereum/log
- swapLog --> github.com/ethersphere/swarm/swap/log

Maybe there's another name that's more appropriate.
It could also be log, but I think it's a bit contrived.

File: swap/swap_test.go [hunk 20]
```diff
@@ -1775,6 +1777,13 @@ func TestSwapLogToFile(t *testing.T) {
 	// now simulate sending the cheque to the creditor from the debitor
 	creditor.sendCheque()
 
+	swapLog.Info("Test")
+	swapLog.SetLogAction(l.Action("disconnecting"))
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap_test.go]: This was refactored now it's `swapLog.SetLogAction("disconnecting")`

File: log/log.go [hunk 21]
```diff
@@ -47,6 +62,104 @@ func Trace(msg string, ctx ...interface{}) {
 	l.Output(msg, l.LvlTrace, CallDepth, ctx...)
 }
 
+func wrapCtx(sl SwapLogger, ctx ...interface{}) []interface{} {
+	for _, elem := range ctx {
+		if elem == "action" && len(ctx)%2 == 0 {
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: This checks if the ctx is even, this avoids the error of unbalanced key value pairs

```
//ctx needs to be even because it's a series of key/value pairs
// no one wants to check for errors on logging functions,
// so instead of erroring on bad input, we'll just make sure
// that things are the right length and users can fix bugs
// when they see the output looks wrong
```

File: log/log.go [hunk 22]
```diff
@@ -47,6 +62,104 @@ func Trace(msg string, ctx ...interface{}) {
...
+// Trace TODO REVIEW THIS COMMENT is a convenient alias for log.Warn with stats
+func (sl SwapLogger) Trace(msg string, ctx ...interface{}) {
+	ctx = wrapCtx(sl, ctx...)
+	sl.logger.Trace(msg, ctx...)
+}
+
+// SetLogAction set the current log action prefix
+func (sl *SwapLogger) SetLogAction(action Action) {
+	//Adds default action *
+	if action == "" {
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: I see your point, when this used actions it made more sense, but now it's a string.
I Agree the SetLogAction should never be called empty.

File: log/log.go [hunk 23]
```diff
@@ -47,6 +62,104 @@ func Trace(msg string, ctx ...interface{}) {
...
+		sl.action = DefaultAction
+		return
+	}
+	//Todo validate it's a specific action, if not default
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: Removed

File: swap/log.go (hunk 11 above)
Comment (santicomp2014) [reply swap/log.go]: @mortelli I refactored it as it's more concise as @Eknir stated.
Thanks for the observations.

File: swap/log.go (hunk 15 above)
Comment (santicomp2014) [reply swap/log.go:12]: Added more information as to what this Logger is used for and it's action field.

File: swap/log.go (hunk 16 above)
Comment (santicomp2014) [reply swap/log.go]: Removed, as it was refactored

File: swap/log.go (hunk 17 above)
Comment (santicomp2014) [reply swap/log.go]: Deprecated and removed in refactor.

File: swap/swap_test.go [hunk 24]
```diff
@@ -1722,12 +1722,14 @@ func TestSwapLogToFile(t *testing.T) {
 
 	// set the log dir to the params
 	params := newDefaultParams(t)
+	params2 := newDefaultParams(t)
 	params.LogPath = logDirDebitor
+	params2.LogPath = logDirDebitor
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap_test.go]: Removed, as creditor.sendCheque() invoques the sentCheque log.
I was going to use it, but it's not necessary.

File: swap/swap_test.go [hunk 25]
```diff
@@ -1722,12 +1722,14 @@ func TestSwapLogToFile(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap_test.go:1808]: I split the test, there is an overlap but it's tolerable.

File: swap/swap_test.go [hunk 26]
```diff
@@ -85,7 +86,6 @@ type swapTestBackend struct {
 func init() {
 	testutil.Init()
 	mrand.Seed(time.Now().UnixNano())
-	swapLog = log.Root()
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap_test.go:102]: swapLog is used with newSwapLogger
This is initialized before use, as it can vary between test.
It should be defined explicitly before using it.

File: swap/swap.go [hunk 27]
```diff
@@ -655,12 +659,13 @@ func (s *Swap) Deploy(ctx context.Context, initialDepositAmount uint64) (contrac
 	// initial topup value
 	opts.Value = big.NewInt(int64(initialDepositAmount))
 	opts.Context = ctx
-	swapLog.Info("Deploying new swap", "owner", opts.From.Hex(), "deposit", opts.Value)
+	swapLog.Info("Deploying new swap", "owner", opts.From.Hex(), "deposit", opts.Value, "action", "deploy")
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap.go]: There are intertwined, I think it's best if `swapLog.SetLogAction("swap_deploy_contract")` is placed before `"Deploy new swap"`

> ...which contract we're talking about, since we're about to have more than 1...

I'm not sure how can we distinguish them, is there a field that can be used for this purpose.

File: swap/swap.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/swap.go]: since this is `go-ethereum/log`, i would just call this `log`. 

to me, `swarmLog` does not seem like an appropriate name for this import, as there is nothing related to swarm in said namespace.

as this is a naming issue, more opinions are welcome

File: swap/swap_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/swap_test.go]: sorry, i probably need to clear this up.

my question was if we could possibly check for log actions being registered after calling swap functions (such as updating a balance, sending a cheque, etc.)

if not, i would just go ahead and remove all the previous setup (dual swaps with keys) as it adds no value when just testing isolated writes.

File: swap/swap_test.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/swap_test.go:102]: got it, thanks

File: swap/swap.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/swap.go]: of course, 1 action for both of these functions seems OK to me :+1: 

regarding distinguishing this contract from others: this one is a chequebook contract—it would be different from, say, an oracle contract.

besides that: remove `swap` from the set action

File: log/log.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (mortelli) [reply log/log.go]: is this not already checked by the `normalize` function though?

taking `Info` as an example, this would be reached through:

(`swap/log`) `Info` → (`go-ethereum/log/logger`) `Info` → `write` →  `newContext` → `normalize`

right?

File: swap/swap_test.go [hunk 28]
```diff
@@ -1795,9 +1795,81 @@ func TestSwapLogToFile(t *testing.T) {
...
+	defer os.RemoveAll(logDirDebitor)
+
+	// set the log dir to the params
+	params := newDefaultParams(t)
+	params.LogPath = logDirDebitor
+
+	testBackend := newTestBackend()
+	defer testBackend.Close()
+	// create test swap
+	testSwap, storeTestSwap := newBaseTestSwapWithParams(t, beneficiaryKey, params, testBackend)
```
Comment (mortelli) [inline swap/swap_test.go]: you aren't doing anything with `storeTestSwap` except removing it later.

i can't see a need for calling `newBaseTestSwapWithParams`, use `newTestSwap` instead, and simplify this test

File: swap/swap_test.go [hunk 29]
```diff
@@ -1795,9 +1795,81 @@ func TestSwapLogToFile(t *testing.T) {
...
 	}
 	logString := string(b)
+
 	if !strings.Contains(logString, "sending cheque") {
 		t.Fatalf("expected the log to contain \"sending cheque\"")
 	}
+
+}
+
+func TestSwapActions(t *testing.T) {
```
Comment (mortelli) [inline swap/swap_test.go:1892]: it's great that you added this, but try to extract the duplicated code between this function and the original `TestSwapLogToFile` test

**Review by mortelli — COMMENTED** [review]
some more feedback (see also the previous unresolved threads for a couple more comments)

File: swap/swap.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap.go]: I switched it to log.
Naming is hard, but I think we can agree l was not a good name, log is generic but appropriate for this use.

If nobody disagrees I will keep it with log.

File: swap/swap_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap_test.go]: Removed this from the test, it's now been migrated to `TestSwapActions`

File: swap/swap.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap.go]: Fixed

File: log/log.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (santicomp2014) [reply log/log.go]: I will remove it, correct It is checked in normalize, but this adds the exception in the context.
I wanted to avoid polluting the log, but it will be the responsibility of the programmer to add the key pair accordingly.

File: swap/log.go [hunk 30]
```diff
@@ -0,0 +1,93 @@
...
+	swapLogger = Logger{
+		action: DefaultAction,
+	}
+	swapLogger.logger = log.New(ctx...)
+	setLoggerHandler(logPath, swapLogger.GetLogger())
+	return swapLogger
+}
+
+// GetLogger returns the underlying logger
+func (sl Logger) GetLogger() (logger log.Logger) {
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/log.go]: This `GetLogger` was used when the code was being exported in the log package.
The only place this is being called is `newLogger` and it is redundant.

Fixed

File: swap/log.go [hunk 31]
```diff
@@ -0,0 +1,93 @@
...
+	return sl.logger
+}
+
+// GetHandler return the Handler assigned to root
+func GetHandler() log.Handler {
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/log.go]: Removed as `GetLogger` same reason.

File: swap/swap.go [hunk 32]
```diff
@@ -655,6 +656,7 @@ func (s *Swap) Deploy(ctx context.Context, initialDepositAmount uint64) (contrac
 	// initial topup value
 	opts.Value = big.NewInt(int64(initialDepositAmount))
 	opts.Context = ctx
+	swapLog.SetLogAction("swap_deploy_contract")
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap.go]: fixed and changed to `deploy_chequebook_contract`

File: swap/swap_test.go (hunk 28 above)
Comment (santicomp2014) [reply swap/swap_test.go]: Fixed --> Simplified with `newTestSwap`

File: swap/swap_test.go (hunk 29 above)
Comment (santicomp2014) [reply swap/swap_test.go:1892]: They are similar, but not necessarily the same.
I would refrain from refactoring until it's used in more logging tests.

File: swap/swap_test.go [hunk 33]
```diff
@@ -1795,9 +1795,81 @@ func TestSwapLogToFile(t *testing.T) {
...
+
+	var b []byte
+	b, err = ioutil.ReadFile(logFile)
+	if err != nil {
+		t.Fatal(err)
+	}
+	logString := string(b)
+
+	if !strings.Contains(logString, "undefined") {
+		t.Fatalf("expected the log to contain \"action undefined\"")
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap_test.go]: fixed

File: swap/swap_test.go [hunk 34]
```diff
@@ -1795,9 +1795,81 @@ func TestSwapLogToFile(t *testing.T) {
...
+	}
+
+	if !strings.Contains(logString, "disconnecting") {
```
[the remark this answers is not in the document]
Comment (santicomp2014) [reply swap/swap_test.go]: fixed

File: swap/swap_test.go (hunk 29 above)
Comment (mortelli) [reply swap/swap_test.go:1892]: yes, at this point only the first lines are the same :+1:

**Review by mortelli — APPROVED** [review]
LGTM.

the only thing i would change about this would be not to use underscores in the swap log actions, because it feels inconsistent with the rest of the log context entries.

and maybe `swap_action` could just be `swap`, but i'm not so sure about that one

File: swap/log.go [hunk 35]
```diff
@@ -0,0 +1,83 @@
+package swap
+
+import (
+	log "github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	// CallDepth is set to 1 in order to influence to reported line number of
+	// the log message with 1 skipped stack frame of calling l.Output()
+	CallDepth = 1
```
Comment (Eknir) [inline swap/log.go]: Where is this variable used? It looks like it is not used anywhere

Comment (santicomp2014) [reply swap/log.go]: Removed, this was used before, when it was part of the log package.
Not used anymore, good point.

File: swap/swap_test.go [hunk 36]
```diff
@@ -1882,9 +1882,72 @@ func TestSwapLogToFile(t *testing.T) {
...
+	swapLog = newSwapLogger(logDirDebitor, swap.params.OverlayAddr)
+
+	swapLog.Info("Test")
+	swapLog.SetLogAction("disconnecting")
+	swapLog.Info("Test")
+	swapLog.Info("Test", "swap_action", "emitting_cheque")
+	swapLog.SetLogAction("sent_cheque")
+	swapLog.Info("Test")
+
+	if logDirDebitor == "" {
```
Comment (Eknir) [inline swap/swap_test.go:1914]: I would test here against `swap.params.LogPath`

**Review by Eknir — COMMENTED** [review]
I left (another review). The biggest concern that I have with these actions is that it will make it more difficult for us (as developers) to use the logger appropriately. We already see that the logger actions are not used correctly during the `init` sequence and I don't know when and why we sometimes append the log with `swap_action`. Apart from this, it is not clear to me where the logger action actually appears (seen from the perspective of a node operator).

I would like to ask feedback from core (@skylenet , for example) about how the need to have this log actions (filter items) is solved elsewhere in the code, because I have not seen that such a log-action is used elsewhere.

File: swap/swap.go [hunk 37]
```diff
@@ -721,6 +722,7 @@ func (s *Swap) Deploy(ctx context.Context, initialDepositAmount uint64) (contrac
 	// initial topup value
 	opts.Value = big.NewInt(int64(initialDepositAmount))
 	opts.Context = ctx
+	swapLog.SetLogAction("deploy_chequebook_contract")
```
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/swap.go:725]: very good catch! :clap:

File: swap/swap.go [hunk 38]
```diff
@@ -155,6 +153,7 @@ func New(dbPath string, prvkey *ecdsa.PrivateKey, backendURL string, params *Par
 	if backendURL == "" {
 		return nil, errors.New("no backend URL given")
 	}
+	swapLog.SetLogAction("init")
```
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/swap.go:156]: you should not see `swaplog` any longer. verify you are running this correctly or else we have a bug (i think the tests should have caught this were it to be the case, but maybe not)

the use you're supposed to make of this change is filtering all log entries down _just_ to the swap-related ones.

File: swap/swap.go [hunk 39]
```diff
@@ -155,6 +153,7 @@ func New(dbPath string, prvkey *ecdsa.PrivateKey, backendURL string, params *Par
...
 	swapLog.Info("connecting to SWAP API", "url", backendURL)
```
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/swap.go:157]: as previously mentioned, `swaplog` should not be showing; the swap action should be showing instead.

the `base` is used in order to identify nodes in case a simulation logs stuff (see #1623)

**Comment by mortelli** [comment]
>We already see that the logger actions are not used correctly during the `init` sequence and I don't know when and why we sometimes append the log with `swap_action`.

using `SetLogAction` does not prevent you from tacking on stuff to the logger context normally, same way you would add anything else (e.g. `honey`, `contract address`, `retry #`, etc.)

now that i see your comment, i do agree that it makes it somewhat confusing if we do it ourselves. i initially thought otherwise due to # of lines of code.

i suggest we set swap actions through the `SetLogAction` function exclusively.

@santicomp2014 thoughts?

File: swap/swap.go [hunk 40]
```diff
@@ -80,17 +80,15 @@ type Params struct {
 }
 
 // newSwapLogger returns a new logger for standard swap logs
-func newSwapLogger(logPath string, overlayAddr []byte) log.Logger {
-	swapLogger := log.New("swaplog", "*", "base", hex.EncodeToString(overlayAddr)[:16])
-	setLoggerHandler(logPath, swapLogger)
-	return swapLogger
+func newSwapLogger(logPath string, overlayAddr []byte) Logger {
```
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/swap.go:83]: a `*Swap`-typed variable does not exist at the point of calling `newSwapLogger`

File: swap/swap.go (hunk 38 above)
[the remark this answers is not in the document]
Comment (Eknir) [reply swap/swap.go:156]: I indeed see this any longer. Probably I built it wrongly.
image

File: swap/swap.go (hunk 37 above)
[the remark this answers is not in the document]
Comment (Eknir) [reply swap/swap.go:725]: I just ran the code and got a different output than expected based on the above. Actually, the output is good! See comment:
https://github.com/ethersphere/swarm/pull/1872#discussion_r344832197 . 
Anyways, even though the output is "correct", I cannot figure out why it is not wrong, as based on just going over the code, I would expect the output to be wrong. 
I just had a discussion with @ralph-pichler on this issue and his suggestion was that the design of having a global state (`logger.action`) be modified by the whole swap-package, possibly different go-routines running simultaneously is not a good design.
One possible solution would be to create a new logger instance for every action.

File: swap/swap.go [hunk 41]
```diff
@@ -710,7 +711,7 @@ func (s *Swap) bindToContractAt(address common.Address) (contract.Contract, erro
 	if err := s.chequebookFactory.VerifyContract(address); err != nil {
 		return nil, fmt.Errorf("contract validation for %v failed: %v", address.Hex(), err)
 	}
-	swapLog.Info("bound to chequebook", "chequebookAddr", address)
+	swapLog.Info("bound to chequebook", "chequebookAddr", address, "swap_action", "bind_contract")
```
Comment (Eknir) [inline swap/swap.go:714]: If we would set `swap_action` with the `logger.SetLogAction`, `swap_action` would appear before `chequebookAddr` and not after, as is now the case. I would like to see it at the same place if we would set it via `logger.SetLogAction`

File: swap/swap.go [hunk 42]
```diff
@@ -80,17 +80,15 @@ type Params struct {
...
+	ctx := []interface{}{"base", hex.EncodeToString(overlayAddr)[:16]}
+	return newLogger(logPath, ctx)
 }
 
 // newPeerLogger returns a new logger for swap logs with peer info
-func newPeerLogger(s *Swap, peerID enode.ID) log.Logger {
-	peerLogger := log.New("swaplog", "*", "base", hex.EncodeToString(s.params.OverlayAddr)[:16], "peer", peerID.String()[:16])
-	setLoggerHandler(s.params.LogPath, peerLogger)
-	return peerLogger
+func newPeerLogger(s *Swap, peerID enode.ID) Logger {
```
[the remark this answers is not in the document]
Comment (Eknir) [reply swap/swap.go:89]: This comment is still relevant

**Comment by santicomp2014** [comment]
Drifted too much from master, the best thing is to create the PR again with the improvements and comments mentioned in this PR
